### ISSUE TITLE
SMACK plugins

### ISSUE THREAD

**Issue by pdeligia** [smackers/smack#128]
Hi guys,

Moving here a discussion that I started privately with Zvonimir. As we are in the process of integrating my Boogie-based lockset analysis with SMACK for the SVComp concurrency benchmarks, it would be nice to have an official way to plug-in add-on tools into SMACK.

A SMACK add-on would be an executable tool with a cleanly defined _interface_ by the SMACK team (e.g. you invoke it from command line, it accepts a bpl file and outputs a bpl file) that enhances the SMACK infrastructure with a new analysis (e.g. lockset analysis). An add-on should be _optional_ so the regular SMACK user does not need to install it, and the core SMACK team does not need to support it. The SMACK team should _only_ support and maintain the SMACK add-on interface and plug-in infrastructure. For example this is how my lockset analysis add-on tool works:

```
.\lockpwn.exe ${PROGRAM}.bpl /o ${OUTPUT}.bpl
```

One suggestion would be to have such add-ons execute after the SMACK llvm2boogie translation has finished, and before SMACK invokes a verifier (Boogie, Corral, etc). Maybe a configuration file could accept the path (and command line arguments) of the available add-ons, and SMACK then invokes the add-ons one after the other in a specified order from the core scripts. This could be also done via command line, but maybe the configuration file option will be more user-friendly and less verbose.

**Comment by michael-emmi** [comment]
Currently this is achieved by first translating only (`-t`) then verifying only (`.bpl` input file):
````
smack.py -t PROGRAM.c OPTIONS
lockpwn.exe a.bpl /o PROCESSED.bpl
smack.py PROCESSED.bpl OPTIONS
````
(Just making sure we’re all aware of that.)

The proposed interface then amounts to simplifying that to:
````
smack.py PROGRAM.c OPTIONS --bpl-processing lockpwn.exe
````
That seems like a harmless enough extension, though one does wonder whether such use cases wouldn’t end up with their own top-level SMACK-wrapping scripts anyways, which could just as easily use the original sequence of 3 commands.

Are there any realistic use cases where you might need a configuration file because there are so many post-processing commands to add? I know of only one other Boogie translator at the moment.

Furthermore, I guess that most Boogie code processors would generally require some command line options, so you would need a way to pass those into `smack.py` as well. Would this end up being any cleaner than a top-level script wrapping `smack.py`?

**Comment by zvonimir** [comment]
Our use case is again primarily SVCOMP.
Last year we had a wrapper script called `smack-svcomp.py` that implemented functionalities similar to what is mentioned here. In the end, I thought that going with a separate SVCOMP wrapper script was not a good idea - it was hacky and it also fell out of sync with `smack.py` soon after the competition was done.
Hence, I've been a proponent of introducing the `--svcomp` command line argument to `smack.py` intead of having a separate script. I told Pantazis to create this item so that we can discuss how to properly support this usage mode.
One thing to note is the Monty's pthreads branch added `--pthread` command line argument to `smack.py`, and maybe we could think of way of support lockpwn through that. I am open to suggestions.

**Comment by pdeligia** [comment]
True the current use case is primarily SVCOMP, I was just trying to think if we could generalise support for add-on tools, but maybe this is not necessary until (and if) more use-cases appear.

Regarding support for lockpwn, it could be a great idea to enable it via the `--pthread` option. Maybe `smack.py` can check if lockpwn is available in some pre-defined location (e.g. could have a `lockpwn` bash script in the `\bin` directory). If the tool is found installed, then it will be invoked after SMACK llvm2bpl. Note that lockpwn requires corral, and also requires corral to run with the flag `/cooperative`, but that should be easy to arrange. If the tool is not found installed, then SMACK proceeds with the verifier as normally.

**Comment by shaobo-he** [comment]
Closed via ce2127c2938f268c19a2a6eceb2bfcb45ffd866f. This is done via the `--transform-bpl` flag.
